bin/utest/signal.c

```diff
@@ -295,6 +296,93 @@ TEST_ADD(signal_sigsuspend) {
   return 0;
 }
 
+/* ======= signal_sigtimedwait ======= */
+int sigtimedwait_child(void *arg) {
+  pid_t ppid = getppid();
+  sched_yield();
+  syscall_ok(kill(ppid, SIGUSR1));
+  return 0;
+}
+
+TEST_ADD(signal_sigtimedwait) {
+  syscall_ok(signal(SIGCONT, sigcont_handler));
+  sigset_t set, current, waitset;
+  __sigemptyset(&set);
+  __sigaddset(&set, SIGUSR1);
+  __sigaddset(&set, SIGCONT);
+  syscall_ok(sigprocmask(SIG_SETMASK, &set, NULL));
+
+  utest_spawn(sigtimedwait_child, NULL);
+
+  siginfo_t info;
+  __sigemptyset(&waitset);
+  __sigaddset(&waitset, SIGUSR1);
+  assert(sigtimedwait(&waitset, &info, NULL) == SIGUSR1);
+  assert(info.si_signo == SIGUSR1);
+
+  syscall_ok(sigprocmask(SIG_BLOCK, NULL, &current));
+  assert(__sigsetequal(&set, &current));
+
+  utest_child_exited(0);
+  return 0;
+}
+
+/* ======= signal_sigtimedwait_timeout ======= */
+static int sigusr2_handled = 0;
+
+void sigtimedwait_timeout_sigusr2_handler(int signo) {
+  sigusr2_handled = 1;
+}
+
+int sigtimedwait_timeout_child(void *arg) {
+  ppid = getppid();
+  while (!sigusr2_handled) {
+    kill(ppid, SIGCONT);
+  }
+  return 0;
+}
+
+TEST_ADD(signal_sigtimedwait_timeout) {
+  syscall_ok(signal(SIGCONT, sigcont_handler));
+  syscall_ok(signal(SIGUSR2, sigtimedwait_timeout_sigusr2_handler));
+  sigset_t set, waitset;
+  __sigemptyset(&set);
+  __sigaddset(&set, SIGUSR1);
+  syscall_ok(sigprocmask(SIG_SETMASK, &set, NULL));
+
+  pid_t cpid = utest_spawn(sigtimedwait_timeout_child, NULL);
+
+  siginfo_t info;
+  __sigemptyset(&waitset);
+  __sigaddset(&waitset, SIGUSR1);
+  timespec_t timeout = {
+    .tv_nsec = -1,
+    .tv_sec = -1,
+  };
+
+  /* tv_nsec is invalid. */
+  syscall_fail(sigtimedwait(&waitset, &info, &timeout), EINVAL);
+
+  /* tv_nsec is valid, but tv_sec < 0. */
+  timeout.tv_nsec = 10000000;
+  syscall_fail(sigtimedwait(&waitset, &info, &timeout), EAGAIN);
+
+  /* Timeout is valid, should be interrupted. */
+  timeout.tv_sec = 0;
+  syscall_fail(sigtimedwait(&waitset, &info, &timeout), EINTR);
+
+  /* After the signal we should be interrupted only once, so next call to
+   * sigtimedwait will timeout. */
+  syscall_ok(kill(cpid, SIGUSR2));
+
+  /* This may be woken up by a timeout or by an interrupt. */
+  nanosleep(&timeout, NULL);
+  syscall_fail(sigtimedwait(&waitset, &info, &timeout), EAGAIN);
```

Comment:
I suggest extracting it into separate test case. You won't have to ensure that child stops sending us signals. We will be 100% sure that nothing strange happens.

Reply:
Good point.